src/harness/testcases/WINNF_FT_S_QPR_testcase.py

```diff
@@ -491,3 +494,342 @@ def test_WINNF_FT_S_QPR_8(self, config_filename):
+    }
+    writeConfig(filename, config)
+
+  @configurable_testcase(generate_QPR_5_default_config)
+  def test_WINNF_FT_S_QPR_5(self, config_filename):
+    """[Configurable] Unsuccessful Grant Request from CBSDs within Coordination
+    Area around Table Mountain Quiet Zone (QZ) with Multiple Grants.
+    """
+    config = loadConfig(config_filename)
+    self.assertValidConfig(
+        config, {
+            'registrationRequests': list,
+            'conditionalRegistrationData': list,
+            'grantRequestsN1': list,
+            'grantRequestsN2': list
+        })
+    self.assertEqual(
+        len(config['registrationRequests']), len(config['grantRequestsN1']))
+    self.assertEqual(
+        len(config['grantRequestsN1']), len(config['grantRequestsN2']))
+
+    # Whitelist FCC ID and User ID.
+    for device in config['registrationRequests']:
+      self._sas_admin.InjectFccId({'fccId': device['fccId']})
+      self._sas_admin.InjectUserId({'userId': device['userId']})
+
+    # Pre-load conditional registration data.
+    if config['conditionalRegistrationData']:
+      self._sas_admin.PreloadRegistrationData({
+          'registrationData': config['conditionalRegistrationData']
+      })
+
+    # Step 1: Register CBSDs.
+    request = {'registrationRequest': config['registrationRequests']}
+    responses = self._sas.Registration(request)['registrationResponse']
+
+    # Check registration response.
+    self.assertEqual(len(responses), len(config['registrationRequests']))
+    grant_request_n1 = []
+    grant_request_n2 = []
+    successful_reg_requests = []
+    for i, response in enumerate(responses):
+      if response['response']['responseCode'] == 0:
+        self.assertTrue('cbsdId' in response)
+        successful_reg_requests.append(config['registrationRequests'][i])
+        config['grantRequestsN1'][i]['cbsdId'] = response['cbsdId']
+        grant_request_n1.append(config['grantRequestsN1'][i])
+        config['grantRequestsN2'][i]['cbsdId'] = response['cbsdId']
+        grant_request_n2.append(config['grantRequestsN2'][i])
+    del request, responses
+
+    if not successful_reg_requests:
+      return  # SAS passes immediately, since none of the registration requests
+      # succeeded in this case.
+
+    # Step 2: For CBSDs successfully registered in Step 1, Send grant request 1
+    request1 = {'grantRequest': grant_request_n1}
+    grant_responses1 = self._sas.Grant(request1)['grantResponse']
+    # Check grant response 1
+    self.assertEqual(len(grant_responses1), len(grant_request_n1))
+
+    # Step 3: For CBSDs successfully registered in Step 1, Send grant request 2
+    request2 = {'grantRequest': grant_request_n2}
+    grant_responses2 = self._sas.Grant(request2)['grantResponse']
+    # Check grant response 2
+    self.assertEqual(len(grant_responses2), len(grant_request_n2))
+
+    for i, device in enumerate(successful_reg_requests):
```

Comment:
Since we are not doing any distance checks, we will be checking for all CBSDs that have been registered and have  approved grants. This is ok for small number of CBSDs but for large numbers, will this be an issue?

Reply:
I think we expect only small numbers of CBSDs in this test case, so I'm not too worried. If you see an obvious way to make it more efficient, please let us know, but I don't expect it to be a problem.